### PR TITLE
feat: complete gRPC server implementation

### PR DESCRIPTION
**Pull request by coryan**
Part of the changes for #58, the only thing missing is a REST API to start the gRPC server.

### CODE REVIEW

File: tests/test_grpc_server.py [hunk 1]
```diff
@@ -71,6 +73,263 @@ def test_read_object(self):
...
+        )
+        self.assertIsNotNone(start.upload_id)
+
+        def streamer():
+            media = TestGrpc._create_block(517 * 1024).encode("utf-8")
+            id = start.upload_id
+            step = storage_pb2.ServiceConstants.Values.MAX_READ_CHUNK_BYTES
+            for offset in range(0, len(media), step):
+                upload_id = id
+                id = None
```
Comment (cojenco) [inline tests/test_grpc_server.py:175]: What is the use for `id`?

File: tests/test_grpc_server.py [hunk 2]
```diff
@@ -71,6 +73,263 @@ def test_read_object(self):
...
+            upload_id=start.upload_id,
+            write_offset=offset,
+            checksummed_data=storage_pb2.ChecksummedData(
+                content=content, crc32c=crc32c.crc32c(content)
+            ),
+            finish_write=False,
+        )
+
+        offset = 2 * QUANTUM
+        content = media[QUANTUM:]
```
Comment (cojenco) [inline tests/test_grpc_server.py:242]:
```suggestion
        content = media[2 * QUANTUM:]
```

File: tests/test_grpc_server.py (hunk 1 above)
Comment (coryan) [reply tests/test_grpc_server.py:175]: Only the first `WriteObjectRequest` can have a `upload_id` field set.  So, after the first iteration `id` is `None`, and therefore only on the first iteration `upload_id` has a valid value.  Terrible?  Maybe.

Is there a more elegant way to yield one value that is "special" and then yield other values?  Oh... can I yield before entering the loop?

File: tests/test_grpc_server.py (hunk 2 above)
Comment (coryan) [reply tests/test_grpc_server.py]: Thanks!

File: tests/test_grpc_server.py (hunk 1 above)
Comment (cojenco) [reply tests/test_grpc_server.py:175]: Not sure I follow... why is it that we only want the first `WriteObjectRequest` to have a valid `upload_id`?

Comment (coryan) [reply tests/test_grpc_server.py:175]: I think the proto spec requires that:

https://github.com/googleapis/googleapis/blob/68e5bf3045d7b3ef55f473494a0cd4566491f899/google/storage/v2/storage.proto#L222-L232

we may not be enforcing it though.

Comment (coryan) [reply tests/test_grpc_server.py]: I change the code to yield once before entering the loop and added a comment.

File: tests/test_grpc_server.py [hunk 3]
```diff
@@ -71,6 +73,271 @@ def test_read_object(self):
...
+                content = media[offset:end]
+                yield storage_pb2.WriteObjectRequest(
+                    write_offset=offset,
+                    checksummed_data=storage_pb2.ChecksummedData(
+                        content=content, crc32c=crc32c.crc32c(content)
+                    ),
+                    finish_write=(end == len(media)),
+                )
+
+        write = self.grpc.WriteObject(streamer(), "fake-context")
```
Comment (cojenco) [inline tests/test_grpc_server.py:195]: I haven't user generators much, but I think we need a for loop in order to generate (and use) each `WriteObjectRequest`

IIUC, `streamer()` is just a generator object. To yield each `WriteObjectRequest`, we either need to utilize a for loop or call `next()` on the generator object.

File: tests/test_grpc_server.py (hunk 1 above)
Comment (cojenco) [reply tests/test_grpc_server.py]: I see! Thanks for the change. Another question on line 195 :)

File: tests/test_grpc_server.py (hunk 3 above)
Comment (cojenco) [reply tests/test_grpc_server.py:195]: Oh wait does `WriteObject` consume a stream-like generator?

Comment (coryan) [reply tests/test_grpc_server.py:195]: So it works right?  I just tested with:

```py
def foo():
...     yield 1
...     yield 2
...     for i in range(10, 15): yield i
... 
>>> 
>>> 
>>> g = foo()
>>> l = list(g)
>>> print(l)
[1, 2, 10, 11, 12, 13, 14]
```

That is what I expected.

Comment (cojenco) [reply tests/test_grpc_server.py:195]: Yep looks good! Sorry about the confusion 😂

**Review by cojenco — APPROVED** [review]
LGTM, thanks!

File: tests/test_grpc_server.py (hunk 3 above)
Comment (coryan) [reply tests/test_grpc_server.py:195]: Thanks to you.

**Pull request merged**